### ISSUE TITLE
max_patch_per_column and loop control

### ISSUE THREAD

**Issue by rgknox** [E3SM-Project/E3SM#5087]
I started this thread because I thought [max_patch_per_column](https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/main/elm_varpar.F90#L89) is larger than it needs to be and is causing some inefficiency in the land model.  But after reading through where it is currently used, I think the bigger issue is about how it is used and how inefficient it is (unless I'm missing something obvious).

It is used to perform initialization calls to Betr and MPP.  @jinyuntang  @bishtgautam and I defer to their thoughts on how this is used.

Initialization for MPP: 
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/main/elm_initializeMod.F90#L1064

Initialization for Betr: 
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/main/elm_initializeMod.F90#L627

Here is where it is set: 
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/main/elm_varpar.F90#L145

Note in that definition, that numpft is really the number of natural+crop pfts.  If you are running with use_crops you get this:
numpft = 50
numcft = 36

without use_crops: 
numpft = 16
numcft = 2

The number of PFTs you get on the column though, which I believe is what this index is supposed to signal, should be the maximum of:  natpft_size, cft_size , maxpatch_urb

Here is an example of where it is used in the ELM code (note all of its uses, aside from intializing MPP and betr, are for executing the same looping behaviour seen here):
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeophys/SoilWaterMovementMod.F90#L1203-L1208

My interpretation is that we should change this so that we loop from pfti to [pftf](https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/data_types/ColumnType.F90#L45), **inside** the column loop.

Other locations used:
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeophys/SoilFluxesMod.F90#L233-L237
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeophys/SoilTemperatureMod.F90#L1775-L1779
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeochem/FireMod.F90#L303-L307
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeochem/CNPhenologyBeTRMod.F90#L3023-L3029
https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeochem/CarbonIsoFluxMod.F90#L865-L869

Perhaps the next question is asking in these above use cases, what is that logic trying to achieve?  Is the objective to loop through every patch on the column? Or is the object to loop through a filter set of patches on the column? There are more efficient ways to do the latter right?

also tagging @thorntonpe @peterdschwartz

**Comment by jinyun1tang** [comment]
@rgknox I recalled I had the same confusion too, but never tried to solve it systematically even in betr. Using pfti to pftf as loop controls definitely improves the efficiency.

**Comment by rgknox** [comment]
@jinyun1tang , regarding the value that is passed to betr, is it supposed to be the maximum possible number of patches on a column, or something else?

**Comment by jinyun1tang** [comment]
@rgknox , yes it is the maximum possible number of patches on a column, so memory is properly allocated. Now betr only runs over natural land unit, or soil column (bare soil + natural pft).
